sourced/ml/models/id_splitter.py

```diff
@@ -0,0 +1,124 @@
+import logging
+import os
+import string
+from typing import List, Sequence, Tuple
+
+import keras
+from keras import backend
+from keras.preprocessing.sequence import pad_sequences
+from modelforge import Model, register_model
+import numpy
+
+from sourced.ml.algorithms.id_splitter.nn_model import (f1score, precision,
+                                                        recall)
+from sourced.ml.models.license import DEFAULT_LICENSE
+
+
+@register_model
+class IdentifierSplitterNN(Model):
+    """
+    Bi-Directionnal LSTM Model. Splits identifiers without need for a conventional pattern.
+    """
+    NAME = "id_splitter_nn"
+    VENDOR = "source{d}"
+    DESCRIPTION = "Model that contains source code identifier splitter BiLSTM weights."
+    LICENSE = DEFAULT_LICENSE
+
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = None
+
+    def construct(self, model: "keras.models.Model" = None,
+                  session: "tf.Session" = None,
+                  maxlen: int = DEFAULT_MAXLEN,
+                  padding: str = DEFAULT_PADDING,
+                  mapping: dict = DEFAULT_MAPPING) -> "IdentifierSplitterNN":
+        assert model is not None
+
+        if session is not None:
+            backend.tensorflow_backend.set_session(session)
+        os.environ["TF_CPP_MIN_LOG_LEVEL"] = "1"
+
+        self._maxlen = maxlen
+        self._padding = padding
+        self._mapping = mapping
+        self._model = model
+        return self
+
+    @property
+    def model(self) -> "keras.models.Model":
+        """
+        Returns the wrapped keras model.
+        """
+        return self._model
+
+    def clear_keras(self):
+        """
+        Clears the keras session used to run the model
+        """
+        keras.backend.clear_session()
+
+    def _generate_tree(self) -> dict:
+        return {
+            "config": self._model.get_config(),
+            "weights": self._model.get_weights()
+            }
+
+    def _load_tree(self, tree: dict):
+        model = keras.models.Model.from_config(tree["config"])
+        model.set_weights(tree["weights"])
+        self.construct(model=model)
+
+    def _prepare_single_identifier(self, identifier: str) -> numpy.array:
+        if self._mapping is None:
+            self._mapping = {c: i for i, c in enumerate(string.ascii_lowercase, start=1)}
```

Comment:
ok, I think I get what kind of mapping is it.
It looks like mapping should be the part of the model itself because who knows how I map the characters during training.